rllib/env/multi_agent_episode.py

```diff
@@ -221,14 +221,21 @@ def __init__(
             AgentID, InfiniteLookbackBuffer
         ] = defaultdict(InfiniteLookbackBuffer)
 
-        # In the `MultiAgentEpisode` we need these buffers to keep track of actions,
-        # that happen when an agent got observations and acted, but did not receive
-        # a next observation, yet. In this case we buffer the action, add the rewards,
-        # and record `is_terminated/is_truncated` until the next observation is
-        # received.
-        self._agent_buffered_actions = {}
-        self._agent_buffered_extra_model_outputs = defaultdict(dict)
-        self._agent_buffered_rewards = {}
+        # Create caches for hanging actions/rewards/extra_model_outputs.
+        # When an agent gets an observation (and then sends an action), but does not
+        # receive immediately a next observation, we store the "hanging" action (and
+        # related rewards and extra model outputs) in the caches postfixed w/ `_end`
+        # until the next observation is received.
+        self._hanging_actions_end = {}
+        self._hanging_extra_model_outputs_end = defaultdict(dict)
+        self._hanging_rewards_end = defaultdict(float)
+
+        # In case of a `cut()` or `slice()`, we also need to store the hanging actions,
```

Comment:
Dumb question: Why do we need these extra buffers? Can't we just copy over the `_hanging_actions_end` to the successor? 
When we concatenate them again `self` has these hanging actions and the first observation in the successor episode for the agent will be used to store the now valid transition.

Reply:
Not a dumb question at all!
An example of where it is crucial to have both these chaches:
```
obs = [
   a0: 0, a1: 0   # ts=0
   a0: 1     # ts=1
   a0: 2   # ts=2
]
```

a1's action at ts=0 now is hanging, b/c there is no follow-up observation after its first one. Thus, when we `cut()` (or slice), we put this hanging action in the "before" buffer:

```
successor = MAEps(
   a0: 2,  <- first observation

   self._hanging_actions_before[a1] = [a1's hanging action]
)
```

Then we continue adding timesteps to the succesor chunk (the one returned by `cut()`). Let's assume that in the successor, eventually, a new obs comes in for a1.
```
successor = MAEps(
   a0: 2,  <- first observation
   a0: 3,
   ...
   a0: 5, a1: 5

   self._hanging_actions_before[a1] = [a1's hanging action]
)
```

At this point, we can't do anything with the hanging action (before) as belongs to an observation (the original one of a1 at ts=0 in the previous chunk) that is NOT part of the successor anymore. It's long gone, so to speak. We thus simply leave this hanging action (before) as-is and ONLY use it to check against a preceeding chunk when we do `concat_episodes()`.

Furthermore, assume that we continue the episode and there is another sequence of missing observations for a1, in which case we also now need to store the hanging action at the end for the time when the next obs for a1 DOES come in. However this hanging action at the end should not be confused with the hanging action at the beginning, they are from different timesteps and have nothing to do with each other.